### DISCUSSION TITLE
Polytype discussion: using it as a parameter and some issues that causes

### DISCUSSION THREAD

**Discussion by apaletta3** [nasa/fprime#1924] [Q&A]
Hey devs,

For the VISORS formation flying mission we're working on here at Georgia Tech, our FPrime deployment is taking in telemetry from outside the deployment (from spacecraft subsystems and from the remote spacecraft) and storing it in the `PolyDb` component for monitoring in one central location. My FaultDetection component is taking in all this telemetry and diagnosing certain fault conditions by comparing the live telemetry values to stored comparison values to see if they are out of bounds.

However, instead of hardcoding these comparison values, we would like them to be reconfigurable in flight so we are storing them as a parameter array with each element in the array corresponding to the comparison value for each telemetry channel. The telemetry channels we are monitoring have multiple data types (int, uint, bool, float), so in order to store comparison values with multiple data types in a single parameter array, we are storing them as `PolyTypes` and directly comparing those to the live telemetry values coming out of the `PolyDb` as `Polytypes` since the `Polytype` conveniently supports those logical comparisons between matching data types.

This is working well for us so far, but I have come across two minor issues that I have questions about:
1. Since `PolyType` is defined as an abstract type in the fpp model (like this: `type Fw.PolyType`), I can't define a default value for the parameter that is an array of polytypes. This is annoying when doing `FW_ASSERT(valid.e == DEFAULT || valid.e == INITIALIZED)` checks when loading this parameter into my component since these will assert unless parameters are properly loaded in the deployment. Do you know if there's a way to specify a default value within the component in cpp form during the deployment intitialization and setup if its not possible in fpp form?

2. Similarly, since the `PolyType` is an abstract type in fpp, it doesn't seem to have an XML representation (please correct me if I'm wrong about this), so it will not be included in the deployment dictionary for the gds to read unless added in manually (it needs to be added in manually otherwise the gds won't start). Do you have any suggestions for how to address this issue? Additionally, are there any plans to better incorporate the polytype into fpp in the future as it has proven to be a very useful construct for telemetry monitoring for us?

**Comment by LeStarch** [comment]
@bocchino would you take a moment to consider these questions?  I think you have more FPP experience than I do.

**Reply by bocchino** [reply]
Sure, here are my thoughts:
1. The most straightforward way to represent the comparison values in F Prime is to have a separate parameter of appropriate type for each comparison value.
2. In theory you could store all the values in one parameter whose type is an FPP struct. This struct can have one field of appropriate type for each comparison value. However, the GDS currently does not support the use of structs as parameter types. See https://fprime-community.github.io/fpp/fpp-users-guide.html#Defining-Components_Parameters_Basic-Parameters. We are working on this issue.
3. PolyType is a C++ union type that is intended for use internally in FSW. Union types are not supported by FPP or the GDS.
